File: solution/py/maxSlidingWindow.py
import collections


class Solution:
    def maxSlidingWindow(self, nums, k):
        res = []
        q = collections.deque()
        
        
        l = 0
        for h in range(len(nums)):
            while q and nums[q[-1]] < nums[h]:
                q.pop()
            q.append(h)


            if l > q[0]:
                q.appendleft()

            if h + 1 >= k:
                res.append(nums[q[0]])
                l += 1
            r += 1
        return res



        
        

File: solution/java/maxSlidingWindow.java
import java.util.*;

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> q = new ArrayDeque<>();
        int[] res = new int[nums.length - k + 1];

        for (int l = 0, h = 0; h < nums.length; h++) {
            while (!q.isEmpty() && nums[q.peekLast()] < nums[h]) {
                q.pollLast();
            }

            q.offerLast(h);

            if (q.getFirst() < l) {
                q.pollFirst();
            }

            if (h - l + 1 == k) {
                res[l] = nums[q.peekFirst()];
                l++;
            }
        }

        return res;

    }
}